File: src/core/run/save.ts
// 런 세이브 직렬화 (순수 — IO 없음). RunState/GameState는 Rng 인스턴스만 빼면 JSON-safe.
// Rng는 state(number)만으로 완전 복원 가능(8.3) → {__rng:state} 마커로 치환·복원. 영속화(localStorage)는 web.
import { Rng } from "../rng.ts";
import type { RunState } from "./types.ts";

/** RunState → JSON 문자열. 모든 Rng(run.rng·battle.rng)를 {__rng:state}로 치환. */
export function serializeRun(run: RunState): string {
  return JSON.stringify(run, (_k, v) => (v instanceof Rng ? { __rng: v.state } : v));
}

/** {__rng:n} 마커를 Rng(state=n)로 깊이 복원. */
function reviveRng(o: unknown): unknown {
  if (o === null || typeof o !== "object") return o;
  const rec = o as Record<string, unknown>;
  if (typeof rec.__rng === "number") { const r = new Rng(0); r.state = rec.__rng; return r; }
  if (Array.isArray(o)) { for (let i = 0; i < o.length; i++) o[i] = reviveRng(o[i]); return o; }
  for (const k in rec) rec[k] = reviveRng(rec[k]);
  return o;
}

/** JSON 문자열 → RunState. 파싱/스키마 손상 시 null(세이브 무시). */
export function deserializeRun(json: string): RunState | null {
  try {
    return reviveRng(JSON.parse(json)) as RunState;
  } catch {
    return null;
  }
}

File: engine/spr-core/src/run/save.rs
//! 런 세이브 직렬화 (TS `web/save.ts` + `core/run/save.ts` 대응). RunState 전 트리 serde → JSON.
//! 영속화(localStorage)는 프론트가 담당. 여기선 순수 직렬화/역직렬화 + 손상 폐기(파싱 실패=None).
//! GameState.log은 #[serde(skip)](append-only·복원 시 빈 벡터) → 전투 중에도 안전 저장.
use super::types::RunState;

/// RunState → JSON 문자열(세이브). 실패 시 빈 문자열.
pub fn serialize_run(run: &RunState) -> String {
    serde_json::to_string(run).unwrap_or_default()
}

/// JSON → RunState. 파싱/스키마 손상 시 None(세이브 폐기 → 새로 시작, TS deserializeRun 대응).
pub fn deserialize_run(json: &str) -> Option<RunState> {
    serde_json::from_str(json).ok()
}
